### PR TITLE
Implement u32 shl, shr, eq, neq, not

### PR DESCRIPTION
**Pull request by tnachen**

### CODE REVIEW

File: assembly/src/parsers/u32_ops.rs [hunk 1]
```diff
@@ -30,6 +33,8 @@ pub fn parse_u32assert(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(),
             span_ops.push(Operation::U32split);
             span_ops.push(Operation::Eqz);
             span_ops.push(Operation::Assert);
+            span_ops.push(Operation::Drop);
+            span_ops.push(Operation::Drop);
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:37]: I don't think we need these drops here.

File: assembly/src/parsers/u32_ops.rs [hunk 2]
```diff
@@ -186,19 +191,74 @@ pub fn parse_u32xor(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
...
+/// Translates u32not assembly instruction to VM operation PAD U32OR SWAP U32XOR.
+pub fn parse_u32not(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0 => return Err(AssemblyError::missing_param(op)),
+        1 => {
+            span_ops.push(Operation::Pad);
+            span_ops.push(Operation::U32or);
+            span_ops.push(Operation::Swap);
+            span_ops.push(Operation::U32xor);
+        }
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:203]: `u32or` and `u32xor` are pretty expensive operations - let's try to implement this without them (even if it requires more cycles). I think we can do something like: `push(2^32) swap incr sub` + making sure the original value is a `u32` (but do double check my logic).

File: assembly/src/parsers/u32_ops.rs [hunk 3]
```diff
@@ -186,19 +191,74 @@ pub fn parse_u32xor(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
...
+pub fn parse_u32shl(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0..=1 => return Err(AssemblyError::missing_param(op)),
+        2 => {
+            // Assert the value is u32
+            span_ops.push(Operation::U32split);
+            span_ops.push(Operation::Not);
+            span_ops.push(Operation::Assert);
+
+            match op.parts()[1].parse::<u32>() {
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:220]: We should probably use `parse_int_param()` function from the parent module for argument parsing here.

File: assembly/src/parsers/u32_ops.rs [hunk 4]
```diff
@@ -186,19 +191,74 @@ pub fn parse_u32xor(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
...
+pub fn parse_u32shr(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0..=1 => return Err(AssemblyError::missing_param(op)),
+        2 => {
+            // Assert the value is u32
+            span_ops.push(Operation::U32split);
+            span_ops.push(Operation::Not);
+            span_ops.push(Operation::Assert);
+
+            match op.parts()[1].parse::<u32>() {
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:247]: We should probably use `parse_int_param()` function from the parent module for argument parsing here.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Some comments in-line.

File: assembly/src/parsers/u32_ops.rs [hunk 5]
```diff
@@ -186,19 +189,74 @@ pub fn parse_u32xor(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
...
+
+            match parse_int_param(op, 1, 1, 31) {
+                Ok(x) => {
+                    let base: u64 = 2;
+                    span_ops.push(Operation::Push(BaseElement::new(base.pow(x))));
+                    span_ops.push(Operation::Mul);
+                    span_ops.push(Operation::U32div);
+                    span_ops.push(Operation::Drop);
+                }
+                Err(e) => return Err(e),
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:226]: We can get rid of `match` in this (and other similar blocks) as:

```Rust
let x = parse_int_param(op, 1, 1, 31)?;
span_ops.push(Operation::Push(BaseElement::new(2u64.pow(x))));
span_ops.push(Operation::Mul);
span_ops.push(Operation::U32div);
span_ops.push(Operation::Drop);
```

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! One more comment inline.

File: assembly/src/parsers/u32_ops.rs (hunk 5 above)
Comment (tnachen) [reply assembly/src/parsers/u32_ops.rs]: Ah yes! Still getting familiar with all the rust goodness :)

File: assembly/src/parsers/u32_ops.rs [hunk 6]
```diff
@@ -219,14 +267,59 @@ pub fn parse_u32revb(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(),
...
+        0 => return Err(AssemblyError::missing_param(op)),
+        1 => {
+            // Check first number is u32
+            span_ops.push(Operation::U32split);
+            span_ops.push(Operation::Not);
+            span_ops.push(Operation::Assert);
+
+            span_ops.push(Operation::Swap);
+
+            // Check first number is u32
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:311]: Should probably be "check second number is u32"

File: assembly/src/parsers/u32_ops.rs [hunk 7]
```diff
@@ -219,14 +267,59 @@ pub fn parse_u32revb(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(),
...
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:284]: Should probably be "check second number is u32"

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! A few more comments inline.

File: assembly/src/parsers/u32_ops.rs [hunk 8]
```diff
@@ -186,19 +189,64 @@ pub fn parse_u32xor(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
...
+        2 => {
+            // Assert the value is u32
+            span_ops.push(Operation::U32split);
+            span_ops.push(Operation::Not);
+            span_ops.push(Operation::Assert);
+
+            let x = parse_int_param(op, 1, 1, 31)?;
+            span_ops.push(Operation::Push(BaseElement::new(2u64.pow(x))));
+            span_ops.push(Operation::Mul);
+            span_ops.push(Operation::U32div);
```
[the remark this answers is not in the document]
Comment (tnachen) [reply assembly/src/parsers/u32_ops.rs:221]: Ah yes I think split is right, I was looking to do a mod 2^32 but didn't do it properly here.

File: assembly/src/parsers/u32_ops.rs [hunk 9]
```diff
@@ -186,19 +189,74 @@ pub fn parse_u32xor(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
...
+pub fn parse_u32not(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0 => return Err(AssemblyError::missing_param(op)),
+        1 => {
+            // Assert the value is u32
+            span_ops.push(Operation::U32split);
+            span_ops.push(Operation::Not);
+            span_ops.push(Operation::Assert);
+
+            span_ops.push(Operation::Push(BaseElement::new(2 ^ 32)));
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:204]: I think this should be: `BaseElement::new(2.pow(32))` (`2 ^ 32` is actually doing an XOR).

File: assembly/src/parsers/u32_ops.rs [hunk 10]
```diff
@@ -219,14 +277,61 @@ pub fn parse_u32revb(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(),
...
+            span_ops.push(Operation::Assert);
+
+            span_ops.push(Operation::Swap);
+
+            // Check second number is u32
+            span_ops.push(Operation::U32split);
+            span_ops.push(Operation::Not);
+            span_ops.push(Operation::Assert);
+
+            span_ops.push(Operation::Swap);
```
Comment (bobbinth) [inline assembly/src/parsers/u32_ops.rs:299]: Do we need this `swap`?

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looking good! Just a couple more small comments.

File: assembly/src/parsers/u32_ops.rs (hunk 10 above)
Comment (tnachen) [reply assembly/src/parsers/u32_ops.rs:299]: I was thinking to keep the original order, but if we're poping it anyway I don't think so

File: assembly/src/parsers/u32_ops.rs (hunk 9 above)
Comment (tnachen) [reply assembly/src/parsers/u32_ops.rs:204]: Ah yes!

**Review by bobbinth — APPROVED** [review]
All good! Thank you!

**Pull request merged**